Consolidate the two auxiliary tries to the main NFD trie

hsivonen opened the pull request:
Closes #2362.

Improves throughput from +15% (Greek) to +119% (English) even with the iterator formulation.

GitHub says -6.42 KB for testdata.postcard.

hsivonen commented:
* [ICU4C PR to main](https://github.com/unicode-org/icu/pull/2152)
* [ICU4C PR to maint-71](https://github.com/unicode-org/icu/pull/2153)
* [ICU4C issue](https://unicode-org.atlassian.net/browse/ICU-22115)

echeran commented on components/collator/src/elements.rs:
```diff
@@ -623,8 +670,13 @@ impl Default for NonPrimary {
-/// unnecessary, since `upcoming` is typically very short.
+/// by ccc.
+///
+/// TODO: This struct carries redundant information, and
```
nit (optional): create issue and update as `TODO(#issue)`

hsivonen replied:
Done

echeran commented on components/normalizer/src/lib.rs:
```diff
@@ -629,32 +676,37 @@ where
-                    character: c,
-                    trie_val: DECOMPOSITION_STARTS_WITH_NON_STARTER,
-                });
+        // TODO: Measure if this check is actually an optimization even in the
```
ditto

hsivonen replied:
Done

echeran commented on components/normalizer/src/properties.rs:
```diff
@@ -0,0 +1,402 @@
+//!
+//! Applications should generally use the full normalizers that are
+//! provided at the top level of this crate. However, the APIs in this
+//! module are provided for callers such as HarfBuzz that specifically
```
```suggestion
//! module are provided for callers such as HarfBuzz that specifically
//! want access to the raw canonical composition operation e.g. for use in a
//! glyph-availability-guided custom normalizer.
```

echeran commented on components/normalizer/src/properties.rs:
```diff
@@ -0,0 +1,402 @@
+    }
+
+    /// Look up the canonical combining class for a scalar value
+    /// represented as `u32`. If the argument is outside the scaler
```
```suggestion
    /// represented as `u32`. If the argument is outside the scalar
```

echeran reviewed (approved):
Looks good.  I added very minor suggestions. Feel free to merge and then follow up later on filing issues to match & link to the unadorned TODOs.

echeran commented:
> * [ICU4C PR to main](https://github.com/unicode-org/icu/pull/2152)
> * [ICU4C PR to maint-71](https://github.com/unicode-org/icu/pull/2153)
> * [ICU4C issue](https://unicode-org.atlassian.net/browse/ICU-22115)

Thanks for doing this. FYI, it sounds like in the future, we can just open one ticket per ICU release for ICU4X export data tool updates.  That would allow us to reuse a ticket for multiple PRs within an ICU release without having to explicitly get TC approval for each subsequent PR.

hsivonen commented:
Thanks for the review. I've addressed the review comments here. I still intend to push one more changeset with a CI job update to full datagen zip once the ICU4C change is tagged for export from the maint-71 branch (by @sffc like last time?).

hsivonen commented:
Thank you for the icuexportdata update. This still needs re-approval for GitHub to permit landing.

hsivonen commented:
@echeran , sorry about requesting review yet again. A conflicting change landed on `main`. Could you, please, press the landing button upon approval so that there isn't a time window for another merge conflict?

echeran reviewed (approved):
Sorry about the conflicts.  Yeah, I should have just merged given the volume of things right before 1.0.  I'll merge now.
